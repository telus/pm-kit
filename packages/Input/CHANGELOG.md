# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [1.2.3](https://github.com/telus/pm-kit/compare/@pm-kit/input@1.2.2...@pm-kit/input@1.2.3) (2020-01-31)


### Bug Fixes

* **input:** moved forwarded ref back to input element ([#49](https://github.com/telus/pm-kit/issues/49)) ([6708413](https://github.com/telus/pm-kit/commit/670841369a202940048224df7e5d21e2919091ac))





## [1.2.2](https://github.com/telus/pm-kit/compare/@pm-kit/input@1.2.1...@pm-kit/input@1.2.2) (2020-01-30)


### Bug Fixes

* **input:** fixed generateId encountering undefined error, input type fix ([#48](https://github.com/telus/pm-kit/issues/48)) ([d4f8516](https://github.com/telus/pm-kit/commit/d4f8516b0f51a7841ffd6e14c53fba2e3443fd9d))





## [1.2.1](https://github.com/telus/pm-kit/compare/@pm-kit/input@1.2.0...@pm-kit/input@1.2.1) (2020-01-29)


### Bug Fixes

* **input:** updated to use new feedbackIcon state values ([#46](https://github.com/telus/pm-kit/issues/46)) ([e34c8f1](https://github.com/telus/pm-kit/commit/e34c8f156bef858059d71cba2e6255ed6eae96a0))
* **input:** updated width ([#44](https://github.com/telus/pm-kit/issues/44)) ([346e0ea](https://github.com/telus/pm-kit/commit/346e0ead70263ac94a2afafb57928a85547a0f04))





# [1.2.0](https://github.com/telus/pm-kit/compare/@pm-kit/input@1.1.1...@pm-kit/input@1.2.0) (2020-01-28)


### Features

* **input:** added input of type password ([#42](https://github.com/telus/pm-kit/issues/42)) ([eb31ef6](https://github.com/telus/pm-kit/commit/eb31ef66feb3f22510e380ca6ef232c0921052e8))
* **input:** fixed the eye image for password input ([#43](https://github.com/telus/pm-kit/issues/43)) ([52fd61a](https://github.com/telus/pm-kit/commit/52fd61ac98acc6d69c716b01ae3e3c9681db3d8f))





## [1.1.1](https://github.com/telus/pm-kit/compare/@pm-kit/input@1.1.0...@pm-kit/input@1.1.1) (2020-01-27)


### Bug Fixes

* **input:** fixed props ([#41](https://github.com/telus/pm-kit/issues/41)) ([dd14c5e](https://github.com/telus/pm-kit/commit/dd14c5e079f0dd5527614340a31fe90012499825))





# [1.1.0](https://github.com/telus/pm-kit/compare/@pm-kit/input@1.0.5...@pm-kit/input@1.1.0) (2020-01-24)


### Features

* **input:** updated to use FeedbackIcon and typography ([#37](https://github.com/telus/pm-kit/issues/37)) ([a89e26d](https://github.com/telus/pm-kit/commit/a89e26df0a120cb42cfb0d66ce0f395b501a10a2))





## [1.0.5](https://github.com/telus/pm-kit/compare/@pm-kit/input@1.0.4...@pm-kit/input@1.0.5) (2020-01-22)

**Note:** Version bump only for package @pm-kit/input





## [1.0.4](https://github.com/telus/pm-kit/compare/@pm-kit/input@1.0.3...@pm-kit/input@1.0.4) (2019-12-11)


### Bug Fixes

* **input:** added fade of label when input is disabled ([#16](https://github.com/telus/pm-kit/issues/16)) ([22386c0](https://github.com/telus/pm-kit/commit/22386c09d153e5ad421eadd3649498af27a52264))
* **input:** fixed label spacing, removed 50% width on container ([#15](https://github.com/telus/pm-kit/issues/15)) ([394d0dc](https://github.com/telus/pm-kit/commit/394d0dc504ef4b8a9efb8e78a86c779bb72832f2))





## [1.0.3](https://github.com/telus/pm-kit/compare/@pm-kit/input@1.0.2...@pm-kit/input@1.0.3) (2019-12-10)


### Bug Fixes

* **input:** fixed ref not being set on input element ([6e5bf54](https://github.com/telus/pm-kit/commit/6e5bf54d8ff4bdc399dcdaf156cc8b5b96115ea1))





## [1.0.2](https://github.com/telus/pm-kit/compare/@pm-kit/input@1.0.1...@pm-kit/input@1.0.2) (2019-12-10)


### Bug Fixes

* **other:** fixed npmignore pattern for stories, updated changelogs ([0e02d0c](https://github.com/telus/pm-kit/commit/0e02d0c53b3a88905d51d4a8cc1b7e8f6da939fa))





## [1.0.1](https://github.com/telus/pm-kit/compare/@pm-kit/input@1.0.0...@pm-kit/input@1.0.1) (2019-12-10)

**Note:** Version bump only for package @pm-kit/input

# 1.0.0 (2019-12-09)

### Features

- **css-reset:** added CssReset package ([#14](https://github.com/telus/pm-kit/issues/14)) ([0999eaf](https://github.com/telus/pm-kit/commit/0999eaf7e63c6c3fc7508ca7acf04b6814887567))
