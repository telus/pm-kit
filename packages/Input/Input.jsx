import React, { forwardRef, useState } from 'react'

import { css } from '@emotion/core'
import PropTypes from 'prop-types'

// pm-kit
import FeedbackIcon from '@pm-kit/feedback-icon'
import { red, parkGreen, greyBlue } from '@pm-kit/colours'
import { size, weight } from '@pm-kit/typography'

// functional components
import generateId from '../../shared/utils/generateId/generateId.js'

// images
import hide from '../../shared/png/hide/hide.png'
import show from '../../shared/png/show/show.png'

export const AUTO_COMPLETE_OPTIONS = [
  'off',
  'on',
  'name',
  'honorific-prefix',
  'given-name',
  'additional-name',
  'family-name',
  'honorific-suffix',
  'nickname',
  'email',
  'username',
  'new-password',
  'current-password',
  'one-time-code',
  'organization-title',
  'organization',
  'street-address',
  'address-line1',
  'address-line2',
  'address-line3',
  'address-level4',
  'address-level3',
  'address-level2',
  'address-level1',
  'country',
  'country-name',
  'postal-code',
  'cc-name',
  'cc-given-name',
  'cc-additional-name',
  'cc-family-name',
  'cc-number',
  'cc-exp',
  'cc-exp-month',
  'cc-exp-year',
  'cc-csc',
  'cc-type',
  'transaction-currency',
  'transaction-amount',
  'language',
  'bday',
  'bday-day',
  'bday-month',
  'bday-year',
  'sex',
  'tel',
  'tel-country-code',
  'tel-national',
  'tel-area-code',
  'tel-local',
  'tel-local-prefix',
  'tel-local-suffix',
  'tel-extension',
  'impp',
  'url',
  'photo',
]

export const FEED_BACK_OPTIONS = ['success', 'error', 'waiting']

export const INPUT_MODE_OPTIONS = ['none', 'text', 'decimal', 'numeric', 'tel', 'search', 'email', 'url']

export const LABEL_TYPE_OPTIONS = ['large', 'small', 'hidden']

export const TYPE_OPTIONS = ['text', 'number', 'password', 'email', 'search', 'tel', 'url']

const inputWrapper = css`
  width: 100%;
`

const passwordInputWrapper = css`
  display: flex;
  border: solid 1px ${parkGreen};
  border-radius: 8px;
  background-color: white;
  width: 100%;
  justify-content: space-around;
`

const inputField = css`
  width: 100%;
  padding: 0 16px;
  font-size: ${size.bodyLarge};
  border: 1px solid ${parkGreen};
  border-radius: 8px;
  height: 48px;
  color: ${parkGreen};
  &:disabled {
    cursor: not-allowed;
    opacity: 0.5;
  }
  &::placeholder {
    color: ${greyBlue};
    font-size: ${size.bodyMedium};
  }
`

const passwordInput = css`
  ${inputField};
  padding: 0;
  border: none;
  height: 46px;
  margin-left: 16px;
  &:focus {
    outline: none;
  }
`

const inputFieldWithError = css`
  border-color: ${red};
`

const inputAndFeedbackWrapper = css`
  position: relative;
`

const feedbackIconWrapper = css`
  position: absolute;
  left: calc(100% + 24px);
  top: 50%;
  transform: translateY(-50%);
  display: flex;
`

const labelContainer = css`
  display: flex;
  flex-wrap: wrap;
  align-items: flex-end;
  color: ${parkGreen};
  font-size: ${size.bodySmall};
  font-weight: ${weight.normal};
  margin: 0 0 0.5rem 16px;
  & label {
    margin-right: 0.5rem;
  }
`

const largeLabelContainer = css`
  ${labelContainer}
  font-size: ${size.bodyLarge};
  font-weight: ${weight.bold};
  margin: 0 0 0.5rem 0.2rem;
`

const isDisabled = css`
  opacity: 0.5;
`
const feedbackError = css`
  color: ${red};
  font-weight: ${weight.normal};
  font-size: ${size.bodySmall};
`

const eyeButton = css`
  border: none;
  background: none;
  display: flex;
  justify-content: space-around;
  margin-right: 5px;
  &:focus {
    outline: none;
  }
`

const eyeImage = css`
  width: 32px;
  height: 22px;
  align-self: center;
`

export const Input = ({
  autoComplete,
  disabled,
  disableUnmasking,
  error,
  feedback,
  feedbackicon,
  forwardedRef,
  id,
  inputMode,
  label,
  labelType,
  name,
  onChange,
  required,
  styles,
  type,
  value,
  ...rest
}) => {
  const [display, setDisplay] = useState(false)
  const inputId = generateId(id, name, label)

  const containerArr = [inputWrapper]
  const eyeButtonArr = [eyeButton]
  const feedbackIconWrapperArr = [feedbackIconWrapper]
  const inputAndFeedbackWrapperArr = [inputAndFeedbackWrapper]
  const inputFieldArr = [inputField]
  const labelContainerArr = [labelContainer]
  const passwordInputWrapperArr = [passwordInputWrapper]

  if (labelType === 'large') {
    labelContainerArr.push(largeLabelContainer)
  }

  if (feedback === 'error') {
    if (type === 'password') {
      passwordInputWrapperArr.push(inputFieldWithError)
    } else {
      inputFieldArr.push(inputFieldWithError)
    }
  }

  if (styles) {
    containerArr.push(styles.containerStyle)
    eyeButtonArr.push(styles.eyeButtonStyle)
    feedbackIconWrapperArr.push(styles.feedbackIconStyle)
    inputAndFeedbackWrapperArr.push(styles.inputAndFeedbackWrapperStyle)
    labelContainerArr.push(styles.labelStyle)
    if (type === 'password') {
      passwordInputWrapperArr.push(styles.inputStyle)
    } else {
      inputFieldArr.push(styles.inputStyle)
    }
  }

  /**
   * this is a workaround for a bug in chrome that moves
   * the cursor into a wrong position if prepended with a space
   */
  const handleKeyDown = e => {
    if (type === 'email' && e.key === ' ') {
      e.preventDefault()
    }
  }

  const renderAutoComplete = (type, autoComplete) => {
    if (!autoComplete && (type === 'email' || type === 'url')) {
      console.log('returning type: ', type)
      return type
    }
    return autoComplete
  }

  const renderInputMode = (type, inputMode) => {
    if (inputMode) {
      return inputMode
    }
    if (type === 'number') {
      return 'decimal'
    }
    if (type === 'password') {
      return 'text'
    }
    return type
  }

  const renderLabel = (label, required, disabled) => {
    const labelText = `${label}${required ? true && '*' : ''}`
    return (
      <label css={disabled ? isDisabled : null} htmlFor={inputId.identity()}>
        {labelText}
      </label>
    )
  }

  const showPassword = () => {
    if (!value || value === '') {
      setDisplay(false)
    } else {
      setDisplay(!display)
    }
  }

  return (
    <div css={containerArr}>
      {labelType !== 'hidden' && (
        <div css={labelContainerArr}>
          {label && renderLabel(label, required, disabled)}
          {feedback === 'error' && error && <span css={feedbackError}>{error}</span>}
        </div>
      )}
      <div css={inputAndFeedbackWrapperArr}>
        {type === 'password' ? (
          <div css={passwordInputWrapperArr}>
            <input
              aria-invalid={feedback}
              aria-label={label}
              autoComplete={renderAutoComplete(type, autoComplete)}
              css={passwordInput}
              disabled={disabled}
              id={inputId.identity()}
              inputMode={renderInputMode(type, inputMode)}
              name={name}
              onChange={onChange}
              onKeyDown={handleKeyDown}
              ref={forwardedRef}
              type={!disableUnmasking && display ? 'text' : 'password'}
              value={value}
              {...rest}
            />
            {!disableUnmasking && (
              <button css={eyeButtonArr} onClick={showPassword}>
                <img css={eyeImage} src={display ? hide : show} alt="show password" />
              </button>
            )}
          </div>
        ) : (
          <input
            aria-invalid={feedback === 'error'}
            aria-label={label}
            autoComplete={renderAutoComplete(type, autoComplete)}
            css={inputFieldArr}
            disabled={disabled}
            feedback={feedback}
            id={inputId.identity()}
            inputMode={renderInputMode(type, inputMode)}
            name={name}
            onChange={onChange}
            onKeyDown={handleKeyDown}
            ref={forwardedRef}
            type={type}
            value={value}
            {...rest}
          />
        )}
        {feedbackicon && feedback && (
          <div css={feedbackIconWrapperArr}>
            <FeedbackIcon state={!feedback ? 'disabled' : feedback} size="24px" />
          </div>
        )}
      </div>
    </div>
  )
}

Input.propTypes = {
  /**
   * Provides suggestion for user's input.
   */
  autoComplete: PropTypes.oneOf(AUTO_COMPLETE_OPTIONS),
  /**
   * Specifies if the Input field should be disabled.
   */
  disabled: PropTypes.bool,
  /**
   * Controls whether the option of unmask password be given or not.
   */
  disableUnmasking: PropTypes.bool,
  /**
   * An error message. Should be limited to text and links. See usage criteria for more details.
   */
  error: PropTypes.string,
  /**
   * A feedback state.
   */
  feedback: PropTypes.oneOf(FEED_BACK_OPTIONS),
  /**
   * Control whether to dispaly feedback-icon or not.
   */
  feedbackicon: PropTypes.bool,
  /**
   * Overrides the mobile keyboard specified by input's type attribute.
   */
  inputMode: PropTypes.oneOf(INPUT_MODE_OPTIONS),
  /**
   * The label.
   */
  label: PropTypes.string.isRequired,
  /**
   * The type of label to display.
   */
  labelType: PropTypes.oneOf(LABEL_TYPE_OPTIONS),
  /**
   * The name.
   */
  name: PropTypes.string,
  /**
   * A callback function to handle changes. For input with `value`, onChange is required.
   */
  onChange: PropTypes.func,
  /**
   * Specifies if Input is a required field.
   */
  required: PropTypes.bool,
  /**
   * Customizes the input according to your needs.
   * Accepts an object of styles in the structure below.
   * {
   *  containerStyle:{},
   *  inputStyle: {}
   *  labelStyle:{},
   *  inputAndFeedbackWrapperStyle: {},
   *  eyeButtonStyle: {},
   *  feedbackIconStyle: {},
   * }
   */
  styles: PropTypes.object,
  /**
   * The HTML5 type of the input field.
   */
  type: PropTypes.oneOf(TYPE_OPTIONS),
  /**
   * Use `value` for controlled Inputs. For uncontrolled Inputs, use React's built-in `defaultValue` prop.
   * For input of type `password`, value is required.
   */
  value: PropTypes.string,
}

Input.defaultProps = {
  autoComplete: undefined,
  disabled: false,
  disableUnmasking: false,
  error: undefined,
  feedback: undefined,
  feedbackicon: false,
  forwardedRef: undefined,
  inputMode: undefined,
  labelType: 'small',
  name: undefined,
  onChange: () => {},
  required: false,
  styles: {},
  type: 'text',
}

const InputWithRef = forwardRef((props, ref) => <Input {...props} forwardedRef={ref} />)

export default InputWithRef
