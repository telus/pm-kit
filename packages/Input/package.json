{
  "name": "@pm-kit/input",
  "version": "1.0.5",
  "main": "dist/index.cjs.js",
  "module": "dist/index.es.js",
  "license": "MIT",
  "publishConfig": {
    "access": "public"
  },
  "author": "Public Mobile",
  "engines": {
    "node": ">=8"
  },
  "bugs": {
    "url": "https://github.com/telus/public-mobile-design-system/issues"
  },
  "peerDependencies": {
    "@emotion/core": "^10.0.22",
    "react": "^16.12.0",
    "react-dom": "^16.12.0"
  },
  "dependencies": {
    "@pm-kit/colours": "^1.1.0",
    "@pm-kit/feedback-icon": "^1.0.0",
    "prop-types": "^15.7.2"
  }
}
