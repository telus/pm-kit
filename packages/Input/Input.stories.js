import React, { useRef, useState } from 'react'
import { withKnobs, text, select, boolean, object } from '@storybook/addon-knobs'

// functional components
import Input, {
  AUTO_COMPLETE_OPTIONS,
  FEED_BACK_OPTIONS,
  INPUT_MODE_OPTIONS,
  LABEL_TYPE_OPTIONS,
  TYPE_OPTIONS,
} from './Input.jsx'

import { version } from './package.json'

export default {
  title: 'Design System|Input',
  component: Input,
  parameters: {
    componentSubtitle: `version ${version}`,
  },
  decorators: [withKnobs],
}

export const Default = () => <Input placeholder="Placeholder" label="Label" />

export const Disabled = () => <Input disabled={true} label="Label" placeholder="Disabled" />

export const Error = () => <Input error="(error)" feedback="error" label="Label" placeholder="Placeholder" />

export const ErrorWithIcon = () => (
  <Input error="(error)" feedback="error" feedbackicon label="Label" placeholder="Placeholder" />
)

export const LargeLabel = () => <Input label="Label" labelType="large" placeholder="Placeholder" />

export const NoLabel = () => <Input label="Label" labelType="hidden" placeholder="Placeholder" />

export const Password = () => {
  const inputRef = useRef()
  const [inputValue, setInputValue] = useState('')
  const inputValueChange = e => {
    setInputValue(e.target.value)
  }
  return (
    <Input
      label="Label"
      onChange={inputValueChange}
      placeholder="Placeholder"
      ref={inputRef}
      type="password"
      value={inputValue}
    />
  )
}

export const Playground = () => {
  const inputRef = useRef()
  const [inputValue, setInputValue] = useState('')
  const inputValueChange = e => {
    setInputValue(e.target.value)
  }
  return (
    <Input
      autoComplete={select('Auto Complete', AUTO_COMPLETE_OPTIONS)}
      disabled={boolean('Disable Input', false)}
      disableUnmasking={boolean('Disable unmasking Password', false)}
      error={text('Error Message', 'Error')}
      feedback={select('Feedback', FEED_BACK_OPTIONS)}
      feedbackicon={boolean('Feedback Icon', false)}
      inputMode={select('Input mode', INPUT_MODE_OPTIONS, undefined)}
      label={text('Label', 'Label')}
      labelType={select('Label Type', LABEL_TYPE_OPTIONS, 'small')}
      onChange={inputValueChange}
      placeholder={text('Placeholder', 'Placeholder')}
      ref={inputRef}
      required={boolean('Required Input', false)}
      styles={object('Styles', {
        containerStyle: {},
        eyeButtonStyle: {},
        feedbackIconStyle: {},
        inputStyle: {},
        inputAndFeedbackWrapperStyle: {},
        labelStyle: {},
      })}
      type={select('Type', TYPE_OPTIONS)}
      value={inputValue}
    />
  )
}

Playground.story = {
  name: 'playground',
  parameters: { docs: { page: null, disable: true } },
}
