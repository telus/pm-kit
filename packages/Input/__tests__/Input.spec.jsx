import React from 'react'
import { render } from 'enzyme'

import Input from '../Input'

describe('Input', () => {
  it('renders', () => {
    const input = render(<Input label="Label" />)

    expect(input).toMatchSnapshot()
  })
})
