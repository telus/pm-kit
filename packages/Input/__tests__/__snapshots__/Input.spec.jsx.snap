// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input renders 1`] = `
.emotion-4 {
  width: 100%;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
  color: #034045;
  font-size: 0.875rem;
  font-weight: 300;
  margin: 0 0 0.5rem 16px;
}

.emotion-1 label {
  margin-right: 0.5rem;
}

.emotion-3 {
  position: relative;
}

.emotion-2 {
  width: 100%;
  padding: 0 16px;
  font-size: 1.125rem;
  border: 1px solid #034045;
  border-radius: 8px;
  height: 48px;
  color: #034045;
}

.emotion-2:disabled {
  cursor: not-allowed;
  opacity: 0.5;
}

.emotion-2::-webkit-input-placeholder {
  color: #819FA2;
  font-size: 1rem;
}

.emotion-2::-moz-placeholder {
  color: #819FA2;
  font-size: 1rem;
}

.emotion-2:-ms-input-placeholder {
  color: #819FA2;
  font-size: 1rem;
}

.emotion-2::placeholder {
  color: #819FA2;
  font-size: 1rem;
}

<div
  class="emotion-4"
>
  <div
    class="emotion-1"
  >
    <label
      class="emotion-0"
      for="label"
    >
      Label
    </label>
  </div>
  <div
    class="emotion-3"
  >
    <input
      aria-invalid="false"
      aria-label="Label"
      class="emotion-2"
      id="label"
      inputmode="text"
      type="text"
    />
  </div>
</div>
`;
